lib/apple_sign_in/user_identity.rb

```diff
@@ -0,0 +1,40 @@
+# frozen_string_literal: true
+
+module AppleSignIn
+  class UserIdentity
+    APPLE_KEY_URL = 'https://appleid.apple.com/auth/keys'
+
+    attr_reader :user_identity, :jwt
+
+    def initialize(user_identity, jwt)
+      @user_identity = user_identity
+      @jwt = jwt
+    end
+
+    def valid?
+      token_data = decoded_jwt
+
+      return true if match_user_identity?(token_data)
+
+      false
+    end
+
+    private
+
+    def decoded_jwt
+      key_hash = apple_key_hash
+      apple_jwk = JWT::JWK.import(key_hash)
+      JWT.decode(jwt, apple_jwk.public_key, true, algorithm: key_hash['alg']).first
+    end
+
+    def match_user_identity?(token_data)
+      user_identity && user_identity == token_data['sub']
+    end
+
+    def apple_key_hash
+      response = Net::HTTP.get(URI.parse(APPLE_KEY_URL))
+      certificate = JSON.parse(response)
+      ActiveSupport::HashWithIndifferentAccess.new(certificate['keys'].first)
```

Comment:
I think that you need to specify the `alg` value to get the correct key from the certificate. Something like this :
```
    header = JSON.parse(Base64.decode64(jwt.split('.').first))
    alg = header['alg']
    kid = header['kid']

    matching_key = certificate['keys'].select{ |h| h['kid'] == kid }
    key_hash = ActiveSupport::HashWithIndifferentAccess.new(matching_key.first)

```

Reply:
why the ```alg```? or did you mean specify the ```kid```?

Reply:
yes, the `kid`

Reply:
Yes, since Apple may have several keys it should find the one with the matching kid instead of using the first one. I don't think searching for a matching alg is necessary though, kid should be unique.
Yet I think that for starters it's fine to use the first certificate and postpone matching the key id until the definition of a specific test for that